File: src/pubsub_cpp/src/sub.cpp
#include <memory> // 스마트 포인터 관련 라이브러리

#include "rclcpp/rclcpp.hpp" // ROS 2 C++ 클라이언트 라이브러리
#include "std_msgs/msg/string.hpp" // ROS 2 표준 메시지 라이브러리
using std::placeholders::_1; // std::bind 사용을 위한 네임스페이스

// 구독자 노드 클래스
class Sub : public rclcpp::Node
{
public:
    Sub() : Node("sub") /** @brief Sub 클래스의 생성자 */
    {
        auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10)); // 메시지 큐의 크기를 10개로 설정한 qos_profile 생성
        subscriber_ = this->create_subscription<std_msgs::msg::String>( // 토픽 이름은 topic, 메시지 큐의 크기는 10개로 설정한 구독자 생성
            "topic", qos_profile, std::bind(&Sub::listener_callback, this, _1)); // topic_callback 함수를 호출
    }

private:
    void listener_callback(const std_msgs::msg::String::SharedPtr msg) const // topic_callback 함수 정의
    {
        RCLCPP_INFO(this->get_logger(), "I heard: '%s'", msg->data.c_str()); // 메시지를 출력
    }
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscriber_; // 구독자 객체
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv); // ROS 2 라이브러리 초기화, 매개변수는 main 함수의 매개변수를 그대로 전달
    auto node = std::make_shared<Sub>(); // Sub 클래스의 객체를 생성
    rclcpp::spin(node); // 노드 실행
    rclcpp::shutdown(); // 노드 종료
    return 0;
}

File: src/pubsub_py/pubsub_py/sub.py
# rclpy 모듈 및 std_msgs모듈 import
import rclpy  # rclpy 모듈 및 std_msgs모듈 import
from rclpy.node import Node
from rclpy.qos import QoSProfile

from std_msgs.msg import String


class Sub(Node):  # subscriber 노드 클래스 Node클래스 상속
    def __init__(self):  # sub 클래스 생성자
        super().__init__('sub')  # super 함수를 통해 Node 클래스의 생성자에 노드 이름을 매게변수로 작성
        qos_profile = QoSProfile(depth=10)
        # String타입의 메시지를 'topic'이름의 토픽으로 subscribe함 ,
        # create_subscription 함수를 사용해 subscriber 생성, 매개변수로는 메시지 타입, 토픽이름,콜백함수, QoS임
        self.subscriber_ = self.create_subscription(
            String, 'topic', self.listener_callback, qos_profile)

    def listener_callback(self, msg):
        self.get_logger().info('I heard: "%s"' % msg.data)


def main(args=None):
    rclpy.init(args=args)  # 주어진 args에 대한 ROS통신 초기화
    node = Sub()  # sub 클래스
    try:
        rclpy.spin(node)  # 프로그램이 종료될 때 까지 큐에 요청된 콜백함수 처리
    except KeyboardInterrupt:
        node.get_logger().info('Keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()  # 노드 종료 시 destroy_node 함수를 통해 subscriber 노드 소멸
        rclpy.shutdown()  # shutdown 함수를 통해 노드 종료

    if __name__ == '__main__':
        main()